packages/block-editor/src/components/block-controls/slot.js

```diff
@@ -13,9 +14,21 @@ import warning from '@wordpress/warning';
  * Internal dependencies
  */
 import groups from './groups';
+import { unlock } from '../../lock-unlock';
+
+const { ComponentsContext } = unlock( privateApis );
 
 export default function BlockControlsSlot( { group = 'default', ...props } ) {
-	const accessibleToolbarState = useContext( ToolbarContext );
+	const toolbarState = useContext( ToolbarContext );
+	const contextState = useContext( ComponentsContext );
+	const fillProps = useMemo(
+		() => [
+			[ ToolbarContext.Provider, { value: toolbarState } ],
+			[ ComponentsContext.Provider, { value: contextState } ],
```

Comment:
Here I did a little change, passing a props object instead of a value for a hard-coded `value` prop. That makes the system more flexible.

We'll want to migrate the `<StyleProvider document={...}>` usage also to this format, and there's a `document` prop instead of `value`.

It would be nice to remove the hard-coded `StyleProvider` from the `Fill` implementation, as added in #38237, and configure the slots and fills that really need this explicitly.

@youknowriad added `StyleProvider` wrappers around specific fills in #31073. There should be eventually migrated to the `fillProps` format.

The `StyleProvider`s from #31073 (specific places) and #38237 (universal for every `<Fill bubblesVirtually>`) often overlap each other, rendering the same provider two times next to each other. I believe the universal one should eventually go away.

Reply:
Thank you for the extra context, Jarda! Your suggestion makes sense in terms of code hygiene and separation of concerns, although there are a few aspects that we should also take into account,

Specifically to `StyleProvider`:
 1. the `emotion` library (and therefore `StyleProvider`) was always meant to be kept as an internal implementation detail to the package. Adopting your suggested approach would require exporting `StyleProvider` from the package. Although maybe we can still keep it as a "generic" style provider (not necessarily Emotion-themed), and if one day we migrate away from Emotion that the provider is not needed anymore, we can leave it around as an empty, no-op context
 2. a good portion of `@wordpress/components` components would need `StyleProvider` to render correctly, meaning that most probably every usage of `Slot` would need to pass `StyleProvider`
 3. related to the point above, removing `StyleProvider` from the `Fill` implementation has the potential to cause visual breaking changes to third-party consumers

Reply:
What is `ComponentsContext`? Can `ToolbarContext` be part of it?

Reply:
`ComponentsContext` is a kind of a generic key-value store, where a parent component can tell child components things like "inside toolbar dropdown menus should be in this toolbar-specific style" or "inside input control the prefix subcomponent should have a certain padding-left". Originally comes from Q and his G2 system.

`ToolbarContext` is a thing coming from Reakit, where a parent component (container) exposes some API for children (items).

Reply:
Yeah, so `ToolbarContext` is basically the same thing as `ComponentsContext` but specific to `Toolbar` components. I wonder if there's a way to unify.